contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,229 @@
+
+            if (isPut) {
+                /**
+                 * Net otoken value for put =
+                 *     (long strike * min (short amount, long amount)) - (short amount * short strike)
+                 */
+                netOtoken = longStrike.mul(FixedPointInt256.min(shortAmount, longAmount)).sub(
+                    shortAmount.mul(shortStrike)
+                );
+            } else {
+                /**
+                 * Net otoken value for call =
+                 *     Min(0, long amount - short amount)
+                 *         - Min(long amount, short amount) * Max(0, long strike - short strike) / long strike)
+                 */
+                if (longStrike == 0) {
+                    // no long otoken
+                    netOtoken = -shortAmount;
+                } else {
+                    netOtoken = FixedPointInt256.min(0, longAmount.sub(shortAmount)).sub(
+                        FixedPointInt256
+                            .min(longAmount, shortAmount)
+                            .mul(FixedPointInt256.max(0, longStrike.sub(shortStrike)))
+                            .div(longStrike)
+                    );
+                }
+            }
+        }
+
+        int256 net = collateralAmount.add(netOtoken);
+        isExcess = net >= 0;
+        netValue = net.intToUint();
+    }
+
+    /**
+     * @dev internal function that ensure each asset type & amout array have the same length, and length <= 1;
+     * @param _vault the vault to check.
+     */
+    function _checkAssetCount(Vault memory _vault) internal pure {
+        // For the currect version, check lengths of short, long, ollateral <= 1.
+        require(_vault.shortOtokens.length <= 1, "MarginCalculator: Too many short otokens in the vault.");
+        require(_vault.longOtokens.length <= 1, "MarginCalculator: Too many long otokens in the vault.");
+        require(_vault.collateralAssets.length <= 1, "MarginCalculator: Too many collateral assets in the vault.");
+
+        require(
+            _vault.shortOtokens.length == _vault.shortAmounts.length,
+            "MarginCalculator: Short asset and amount mismatch."
+        );
+        require(
+            _vault.longOtokens.length == _vault.longAmounts.length,
+            "MarginCalculator: Long asset and amount mismatch."
+        );
+        require(
+            _vault.collateralAssets.length == _vault.collateralAmounts.length,
+            "MarginCalculator: Collateral asset and amount mismatch."
+        );
+    }
+
+    /**
+     * @dev internal function that check the long asset is valid for the short asset
+     * @param _vault the vault to check.
+     */
+    function _checkLongAsset(Vault memory _vault) internal view {
+        if (_vault.longOtokens.length == 0 || _vault.shortOtokens.length == 0) return;
+
+        OtokenInterface long = OtokenInterface(_vault.longOtokens[0]);
+        OtokenInterface short = OtokenInterface(_vault.shortOtokens[0]);
+
```

Comment:
also check that both oTokens are isValidOTokens?

Reply:
@haythem96 and I discussed this and we think it would be better to check that in Controller (mint action, which we already did), otherwise the Calculator need to add Whitelist as dependency